### PR TITLE
Adds support for newlines and carrage returns

### PR DESCRIPTION
**Pull request by tocsoft**
This adds support for newlines in the layout of glyphs.

### CODE REVIEW

File: NOpenType/Renderer.cs [hunk 1]
```diff
@@ -259,22 +259,40 @@ private IEnumerable<GlyphLayout> Layout(int x, int y, string text, TextOptions o
             int scalingFactor = ScalingFactor(options.FontSize);
             int xx = (int)(x * FontToPixelDivisor) / scalingFactor;
             int yy = (int)(y * FontToPixelDivisor) / scalingFactor;
-
-            int drawheightEM = _typeface.Bounds.YMax - _typeface.Bounds.YMin;
-            int lineHeight = (int)(drawheightEM * options.LineHeight);
+            int startXX = xx;
+            
+            // this will need looking at as I don't know if there somethign inside a font that should tell us the real base lineheight, this renders nicely in my tests.
+            int lineHeight = (int)(EmSquare.Size * options.LineHeight);
```
Comment (vidstige) [inline NOpenType/Renderer.cs]: yes, there is a settings for this. Check out the specs. You can probably find it in the source here also if you look through the table-classes.

File: NOpenType/Renderer.cs [hunk 2]
```diff
@@ -259,22 +259,40 @@ private IEnumerable<GlyphLayout> Layout(int x, int y, string text, TextOptions o
...
+                        {
+                            glyph = glyph,
+                            TopLeft = new Point<int>(xx, yy),
+                            BottomRight = new Point<int>(glyphWidth + xx, drawheightEM + yy)
+                        };
 
-                xx += glyphWidth;
+                        xx += glyphWidth;
+                        break;
+                }
```
Comment (vidstige) [inline NOpenType/Renderer.cs:294]: My thought for this code for the future, is to only work with zero-based font unit coordinate system. Now we mix pixel-based and font based. For the future we could call functions in a class to advance xx/yy to get smooth right-left support aswell :-)

File: NOpenType/Renderer.cs (hunk 1 above)
Comment (vidstige) [reply NOpenType/Renderer.cs]: Just fix this and lets merge.

File: NOpenType/TextOptions.cs [hunk 3]
```diff
@@ -26,7 +26,10 @@ public class TextOptions
         /// <value>
         /// The height of the line.
         /// </value>
-        public float LineHeight { get; set; } = 1;
+        /// <remarks>
+        /// Default is 1.25 the standard glyph height as it makes for a nicer line seperation for a default rendering.
+        /// </remarks>
+        public float LineHeight { get; set; } = 1.25f;
```
Comment (vidstige) [inline NOpenType/TextOptions.cs]: keep at 1

File: NRasterizer.CLI/Program.cs [hunk 4]
```diff
@@ -51,8 +52,9 @@ private static void Grayscale(Image image)
 
         private void DrawSoft(FileInfo fontPath, FileInfo target, string text)
         {
+            var lineCount = text.Where(c => c == '\n').Count() + 1;
```
Comment (vidstige) [inline NRasterizer.CLI/Program.cs]: You can just do `text.Count(c => c == '\n') + 1`

**Comment by vidstige** [comment]
:+1: for adding tests.

File: NOpenType/Renderer.cs (hunk 1 above)
Comment (tocsoft) [reply NOpenType/Renderer.cs]: Found the spec for the correct way to calculate this, the required table was missing, so I've now added a `MetricsTable` this maps the the OS\2 table in the spec and passed the correct value down the line into the `Typeface`.

File: NOpenType/TextOptions.cs (hunk 3 above)
Comment (tocsoft) [reply NOpenType/TextOptions.cs]: yeah total agree is should be 1 now I've found out that the standard line spacing includes the extra gap.

File: NRasterizer.CLI/Program.cs (hunk 4 above)
Comment (tocsoft) [reply NRasterizer.CLI/Program.cs]: yeah, I keep forgetting the overloads on things like count actually take a filter in.

**Pull request merged**